File: backend/src/main/java/com/vink/backend/store/VehicleStore.java
package com.vink.backend.store;

import com.vink.backend.model.vehicles.VehicleModels.*;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Java port of server/src/data/vehicleStore.ts.
 * In-memory seeded dataset mirroring the original Express/TS demo store -
 * same field names, same generation logic, same record counts.
 */
@Component
public class VehicleStore {

    private static final double[] SA_BOUNDS = {-34.0, -22.0, 16.0, 33.0}; // latMin,latMax,lngMin,lngMax

    private record City(String name, double lat, double lng) {}

    private static final List<City> CITIES = List.of(
            new City("Johannesburg", -26.2041, 28.0473),
            new City("Cape Town", -33.9249, 18.4241),
            new City("Durban", -29.8587, 31.0218),
            new City("Pretoria", -25.7479, 28.2293),
            new City("Port Elizabeth", -33.9608, 25.6022),
            new City("Bloemfontein", -29.0852, 26.1596),
            new City("East London", -33.0153, 27.9116),
            new City("Polokwane", -23.9045, 29.4689)
    );

    private record MakeModels(String make, List<String> models) {}

    private static final List<MakeModels> MAKES = List.of(
            new MakeModels("Toyota", List.of("Hilux", "Land Cruiser", "Fortuner", "Quantum")),
            new MakeModels("Ford", List.of("Ranger", "Transit", "Everest")),
            new MakeModels("Isuzu", List.of("D-Max", "NLR 150", "NPR")),
            new MakeModels("VW", List.of("Transporter", "Amarok", "Crafter")),
            new MakeModels("Mercedes", List.of("Sprinter", "Vito", "Actros")),
            new MakeModels("Hino", List.of("300", "500", "700"))
    );

    private static final List<String> FLEETS = List.of("Delivery", "Executive", "Field Ops", "Security", "Maintenance");
    private static final List<String> COLOURS = List.of("White", "Silver", "Black", "Blue", "Red", "Orange");
    private static final List<String> STATUSES = List.of("moving", "moving", "moving", "idle", "idle", "stopped", "offline", "maintenance");
    private static final List<String> DRIVER_NAMES = List.of(
            "Sipho Nkosi", "Thabo Dlamini", "Lerato Mokoena", "Nomsa Khumalo", "David van der Berg",
            "Fatima Essop", "Johan Botha", "Priya Naidoo", "Bongani Zulu", "Elaine Mostert",
            "Khulekani Mthembu", "Ayanda Cele", "Pieter du Plessis", "Zanele Mahlangu", "Ravi Pillay",
            "Amina Mohamed", "Chris Steyn", "Nolwazi Dube", "Stefan Kruger", "Mmaphuti Mokgopa",
            "Lungelo Buthelezi", "Sonia Ferreira", "Thandeka Ntuli", "Wayne Jacobs", "Keabetswe Sithole",
            "Lebo Masondo", "Mark Thompson", "Zinhle Ndlovu", "André Fourie", "Busisiwe Mkhize",
            "Sanele Gumbi", "Rebecca Jooste", "Mthokozisi Ndaba", "Charlene van Wyk", "Nkosinathi Mhlanga",
            "Desiré Bosman", "Vusi Shabalala", "Lindi Pietersen", "Siyanda Majola", "Henk Swanepoel"
    );
    private static final List<String> PROVINCES = List.of("GP", "WC", "KZN", "EC", "FS", "LP");
    private static final List<String> COMM = List.of("cellular_4g", "cellular_4g", "cellular_4g", "cellular_3g", "satellite");
    private static final List<String> ALERT_TYPES = List.of("speeding", "geofence_breach", "low_fuel", "harsh_braking", "engine_off", "sos", "tampering", "accident");

    public final List<Vehicle> vehicles = new ArrayList<>();
    public final List<Driver> drivers = new ArrayList<>();
    public final List<Trip> trips = new ArrayList<>();
    public final List<VehicleAlert> alerts = new ArrayList<>();
    public final List<Geofence> geofences = new ArrayList<>();
    public final List<FleetSnapshot> fleetHistory = new ArrayList<>();

    public VehicleStore() {
        seed();
    }

    private static int rand(int min, int max) {
        return ThreadLocalRandom.current().nextInt(min, max + 1);
    }

    private static double randF(double min, double max) {
        double v = ThreadLocalRandom.current().nextDouble(min, max);
        return Math.round(v * 10000.0) / 10000.0;
    }

    private static String ago(long minutes) {
        return Instant.now().minus(minutes, ChronoUnit.MINUTES).toString();
    }

    private static String inFuture(long days) {
        return Instant.now().plus(days, ChronoUnit.DAYS).toString();
    }

    private City nearCity(int i) {
        return CITIES.get(i % CITIES.size());
    }

    private GpsPosition makePosition(int i) {
        City c = nearCity(i);
        GpsPosition p = new GpsPosition();
        p.setLat(Math.round((c.lat() + randF(-0.15, 0.15)) * 1_000_000.0) / 1_000_000.0);
        p.setLng(Math.round((c.lng() + randF(-0.15, 0.15)) * 1_000_000.0) / 1_000_000.0);
        p.setAltitude(rand(10, 1800));
        p.setAccuracy(rand(3, 15));
        p.setHeading(rand(0, 359));
        p.setTimestamp(ago(rand(0, 5)));
        return p;
    }

    private SensorData makeSensors(String status) {
        boolean moving = "moving".equals(status);
        SensorData s = new SensorData();
        s.setSpeedKph(moving ? rand(20, 120) : 0);
        s.setFuelPercent(rand(8, 98));
        s.setEngineTempC(moving ? rand(82, 105) : rand(20, 50));
        s.setEngineOn(!status.equals("offline") && !status.equals("stopped"));
        s.setOdometreKm(rand(5000, 250000));
        s.setRpm(moving ? rand(1200, 3800) : rand(0, 900));
        s.setBatteryV(randF(11.8, 14.4));
        s.setDoorOpen(ThreadLocalRandom.current().nextDouble() < 0.05);
        s.setIgnitionOn(status.equals("moving") || status.equals("idle"));
        return s;
    }

    private void seed() {
        // Vehicles
        for (int i = 0; i < 50; i++) {
            MakeModels mk = MAKES.get(i % MAKES.size());
            String model = mk.models().get(i % mk.models().size());
            String status = STATUSES.get(i % STATUSES.size());

            Vehicle v = new Vehicle();
            v.setId(UUID.randomUUID().toString());
            v.setPlateNumber(PROVINCES.get(i % 6) + " " + rand(100, 999) + " "
                    + (char) ('A' + (i % 26)) + (char) ('A' + ((i + 3) % 26)));
            v.setMake(mk.make());
            v.setModel(model);
            v.setYear(rand(2016, 2024));
            v.setColour(COLOURS.get(i % COLOURS.size()));
            v.setVin("VIN" + String.format("%014d", i));
            v.setFleetGroup(FLEETS.get(i % FLEETS.size()));
            v.setDriverId(!status.equals("offline") ? "DRV-" + String.format("%03d", i) : null);
            v.setStatus(status);
            v.setCommChannel(COMM.get(i % COMM.size()));
            v.setPosition(makePosition(i));
            v.setSensors(makeSensors(status));
            v.setLastSeen(ago(status.equals("offline") ? rand(30, 480) : rand(0, 3)));
            v.setRegisteredAt(ago(rand(1440, 525600)));
            v.setSimIccid("8927" + String.format("%015d", i));
            v.setDeviceId("GT06N-" + String.format("%06d", i));
            v.setInsuranceExpiry(inFuture(rand(30, 365)));
            v.setLicenceExpiry(inFuture(rand(60, 730)));
            v.setNextServiceKm(rand(1000, 10000));
            vehicles.add(v);
        }

        // Drivers
        for (int i = 0; i < 40; i++) {
            Driver d = new Driver();
            d.setId("DRV-" + String.format("%03d", i));
            d.setName(DRIVER_NAMES.get(i));
            d.setLicenceNo("SA" + rand(1000000, 9999999));
            d.setPhone("+2760" + rand(1000000, 9999999));
            d.setEmail("driver" + i + "@vink-fleet.co.za");
            d.setStatus(i % 8 == 0 ? "suspended" : "active");
            d.setAssignedVehicleId(null);
            d.setTotalTrips(rand(50, 1200));
            d.setTotalKm(rand(5000, 180000));
            d.setSafetyScore(rand(55, 100));
            d.setJoinedAt(ago(rand(1440, 525600)));
            String[] parts = d.getName().split(" ");
            StringBuilder initials = new StringBuilder();
            for (String p : parts) if (!p.isEmpty()) initials.append(p.charAt(0));
            d.setAvatarInitials(initials.toString().toUpperCase().substring(0, Math.min(2, initials.length())));
            drivers.add(d);
        }

        // Trips
        for (int i = 0; i < 80; i++) {
            int vehicleIdx = i % 50;
            City cityA = CITIES.get(i % CITIES.size());
            City cityB = CITIES.get((i + 1) % CITIES.size());
            String start = ago(rand(30, 1440));
            boolean ended = i % 6 != 0;
            double dist = randF(5, 420);
            int avgSpd = rand(35, 95);

            List<TripPoint> points = new ArrayList<>();
            Instant startInstant = Instant.parse(start);
            for (int p = 0; p < 8; p++) {
                TripPoint tp = new TripPoint();
                tp.setLat(Math.round((cityA.lat() + (cityB.lat() - cityA.lat()) * (p / 7.0) + randF(-0.02, 0.02)) * 1_000_000.0) / 1_000_000.0);
                tp.setLng(Math.round((cityA.lng() + (cityB.lng() - cityA.lng()) * (p / 7.0) + randF(-0.02, 0.02)) * 1_000_000.0) / 1_000_000.0);
                tp.setSpeedKph(rand(0, 130));
                tp.setTimestamp(startInstant.plusMillis((long) p * rand(180000, 600000)).toString());
                points.add(tp);
            }

            Trip t = new Trip();
            t.setId(UUID.randomUUID().toString());
            t.setVehicleId("VEH-PLACEHOLDER-" + vehicleIdx); // patched below
            t.setDriverId("DRV-" + String.format("%03d", vehicleIdx % 40));

            GpsPosition startPos = new GpsPosition();
            startPos.setLat(cityA.lat()); startPos.setLng(cityA.lng());
            startPos.setAltitude(1200); startPos.setAccuracy(5);
            startPos.setHeading(rand(0, 359)); startPos.setTimestamp(start);
            t.setStartPos(startPos);

            if (ended) {
                GpsPosition endPos = new GpsPosition();
                endPos.setLat(cityB.lat()); endPos.setLng(cityB.lng());
                endPos.setAltitude(1100); endPos.setAccuracy(5);
                endPos.setHeading(rand(0, 359)); endPos.setTimestamp(ago(rand(0, 120)));
                t.setEndPos(endPos);
                t.setEndTime(ago(rand(0, 120)));
            }
            t.setStartTime(start);
            t.setDistanceKm(dist);
            t.setAvgSpeedKph(avgSpd);
            t.setMaxSpeedKph(Math.min(rand(avgSpd + 10, 160), 160));
            t.setFuelUsedL(Math.round(dist * randF(0.06, 0.12) * 100.0) / 100.0);
            t.setPoints(points);
            t.setOngoing(!ended);
            trips.add(t);
        }

        // Alerts
        for (int i = 0; i < 30; i++) {
            String type = ALERT_TYPES.get(i % ALERT_TYPES.size());
            String sev = (type.equals("sos") || type.equals("accident")) ? "critical"
                    : (type.equals("speeding") || type.equals("tampering")) ? "warning" : "info";
            GpsPosition pos = makePosition(i);

            VehicleAlert a = new VehicleAlert();
            a.setId(UUID.randomUUID().toString());
            a.setVehicleId("VEH-PLACEHOLDER-" + (i % 50)); // patched below
            a.setPlateNumber("GP " + rand(100, 999) + " AB");
            a.setDriverId("DRV-" + String.format("%03d", i % 40));
            a.setType(type);
            a.setSeverity(sev);
            a.setMessage(switch (type) {
                case "speeding" -> "Vehicle exceeded speed limit — " + rand(80, 160) + " km/h in 60 zone";
                case "geofence_breach" -> "Vehicle exited authorised geofence boundary";
                case "low_fuel" -> "Fuel level critically low — " + rand(5, 15) + "% remaining";
                case "harsh_braking" -> "Harsh braking event detected at " + rand(60, 110) + " km/h";
                case "engine_off" -> "Engine switched off outside designated area";
                case "sos" -> "SOS panic button activated by driver";
                case "tampering" -> "GPS device tampering detected";
                case "accident" -> "Possible accident — sudden deceleration to 0 km/h";
                default -> "";
            });
            a.setPosition(pos);
            a.setDetectedAt(ago(rand(0, 480)));
            a.setResolvedAt(i % 4 == 0 ? ago(rand(0, 60)) : null);
            a.setAcknowledged(i % 3 == 0);
            a.setValue(type.equals("speeding") ? (double) rand(80, 160) : type.equals("low_fuel") ? (double) rand(5, 15) : null);
            alerts.add(a);
        }

        // Geofences
        geofences.addAll(List.of(
                geofence("Johannesburg HQ", "Head office compound", -26.2041, 28.0473, 500, true, "exit", List.of("Delivery", "Executive"), 43800, 12),
                geofence("Cape Town Depot", "Western Cape warehouse", -33.9249, 18.4241, 800, true, "both", List.of("Delivery"), 20160, 7),
                geofence("Durban Port Zone", "Port operations area", -29.8587, 31.0218, 1200, true, "enter", List.of("Field Ops", "Delivery"), 10080, 3),
                geofence("Pretoria Admin Zone", "Government precinct", -25.7479, 28.2293, 600, true, "exit", List.of("Executive"), 5040, 2),
                geofence("Night Curfew Area", "After-hours restricted", -26.1052, 28.0560, 3000, false, "enter", List.of("Delivery", "Field Ops"), 2880, 0)
        ));

        // Fleet history (24 hourly snapshots)
        for (int i = 0; i < 24; i++) {
            FleetSnapshot fs = new FleetSnapshot();
            fs.setTimestamp(ago((23 - i) * 60L));
            fs.setTotalVehicles(50);
            fs.setMovingNow(rand(18, 32));
            fs.setIdleNow(rand(5, 12));
            fs.setOfflineNow(rand(2, 6));
            fs.setMaintenanceNow(rand(1, 4));
            fs.setActiveAlerts(rand(3, 18));
            fs.setAvgSpeedKph(rand(45, 78));
            fs.setAvgFuelPercent(rand(52, 72));
            fs.setTotalKmToday(rand(800, 4200));
            fs.setActiveDrivers(rand(20, 35));
            fs.setGeofenceBreaches(rand(0, 5));
            fleetHistory.add(fs);
        }

        // Patch trip vehicle ids to real vehicle ids
        for (int i = 0; i < trips.size(); i++) {
            trips.get(i).setVehicleId(vehicles.get(i % vehicles.size()).getId());
        }

        // Patch alert vehicle/driver/plate to real vehicle data
        for (int i = 0; i < alerts.size(); i++) {
            Vehicle v = vehicles.get(i % vehicles.size());
            VehicleAlert a = alerts.get(i);
            a.setVehicleId(v.getId());
            a.setPlateNumber(v.getPlateNumber());
            a.setDriverId(v.getDriverId());
        }
    }

    private Geofence geofence(String name, String description, double lat, double lng, double radius,
                               boolean active, String triggerOn, List<String> groups, long ageMinutes, int breachCount) {
        Geofence g = new Geofence();
        g.setId(UUID.randomUUID().toString());
        g.setName(name);
        g.setDescription(description);
        g.setShape("circle");
        g.setCenterLat(lat);
        g.setCenterLng(lng);
        g.setRadiusM(radius);
        g.setActive(active);
        g.setTriggerOn(triggerOn);
        g.setAlertVehicleGroups(groups);
        g.setCreatedAt(ago(ageMinutes));
        g.setBreachCount(breachCount);
        return g;
    }

    /** Java port of getFleetSnapshot() in vehicleStore.ts */
    public FleetSnapshot getFleetSnapshot() {
        FleetSnapshot fs = new FleetSnapshot();
        fs.setTimestamp(Instant.now().toString());
        fs.setTotalVehicles(vehicles.size());
        fs.setMovingNow(vehicles.stream().filter(v -> v.getStatus().equals("moving")).count());
        fs.setIdleNow(vehicles.stream().filter(v -> v.getStatus().equals("idle")).count());
        fs.setOfflineNow(vehicles.stream().filter(v -> v.getStatus().equals("offline")).count());
        fs.setMaintenanceNow(vehicles.stream().filter(v -> v.getStatus().equals("maintenance")).count());
        fs.setActiveAlerts(alerts.stream().filter(a -> a.getResolvedAt() == null).count());

        List<Vehicle> movingVehicles = vehicles.stream().filter(v -> v.getStatus().equals("moving")).collect(Collectors.toList());
        double avgSpeed = movingVehicles.stream().mapToDouble(v -> v.getSensors().getSpeedKph()).average().orElse(0);
        fs.setAvgSpeedKph(Math.round(avgSpeed * 10.0) / 10.0);

        double avgFuel = vehicles.stream().mapToDouble(v -> v.getSensors().getFuelPercent()).average().orElse(0);
        fs.setAvgFuelPercent(Math.round(avgFuel * 10.0) / 10.0);

        String cutoff = ago(1440);
        double kmToday = trips.stream().filter(t -> t.getStartTime().compareTo(cutoff) > 0).mapToDouble(Trip::getDistanceKm).sum();
        fs.setTotalKmToday(Math.round(kmToday * 10.0) / 10.0);

        long activeDrivers = drivers.stream().filter(d -> d.getStatus().equals("active")
                && vehicles.stream().anyMatch(v -> d.getId().equals(v.getDriverId()) && v.getStatus().equals("moving"))).count();
        fs.setActiveDrivers(activeDrivers);

        long geofenceBreaches = alerts.stream().filter(a -> a.getType().equals("geofence_breach") && a.getResolvedAt() == null).count();
        fs.setGeofenceBreaches(geofenceBreaches);

        return fs;
    }

    public Optional<Vehicle> findVehicle(String id) {
        return vehicles.stream().filter(v -> v.getId().equals(id)).findFirst();
    }

    public Optional<Driver> findDriver(String id) {
        return drivers.stream().filter(d -> d.getId().equals(id)).findFirst();
    }

    public Optional<Geofence> findGeofence(String id) {
        return geofences.stream().filter(g -> g.getId().equals(id)).findFirst();
    }

    public Optional<VehicleAlert> findAlert(String id) {
        return alerts.stream().filter(a -> a.getId().equals(id)).findFirst();
    }
}

File: server/src/data/vehicleStore.ts
import { v4 as uuid } from "uuid";
import type {
  Vehicle, Driver, Trip, VehicleAlert, Geofence, FleetSnapshot,
  GpsPosition, SensorData,
} from "../types/vehicles.js";

const rand  = (min: number, max: number) => Math.floor(Math.random() * (max - min + 1)) + min;
const randF = (min: number, max: number) => +(Math.random() * (max - min) + min).toFixed(4);
const ago   = (min: number)              => new Date(Date.now() - min * 60_000).toISOString();

// South Africa bounding box
const SA_BOUNDS = { latMin: -34.0, latMax: -22.0, lngMin: 16.0, lngMax: 33.0 };

// City clusters for realistic positioning
const CITIES = [
  { name: "Johannesburg", lat: -26.2041, lng: 28.0473 },
  { name: "Cape Town",    lat: -33.9249, lng: 18.4241 },
  { name: "Durban",       lat: -29.8587, lng: 31.0218 },
  { name: "Pretoria",     lat: -25.7479, lng: 28.2293 },
  { name: "Port Elizabeth", lat: -33.9608, lng: 25.6022 },
  { name: "Bloemfontein", lat: -29.0852, lng: 26.1596 },
  { name: "East London",  lat: -33.0153, lng: 27.9116 },
  { name: "Polokwane",    lat: -23.9045, lng: 29.4689 },
];

function nearCity(i: number): { lat: number; lng: number } {
  const city = CITIES[i % CITIES.length];
  return {
    lat: +(city.lat + randF(-0.15, 0.15)).toFixed(6),
    lng: +(city.lng + randF(-0.15, 0.15)).toFixed(6),
  };
}

function makePosition(i: number): GpsPosition {
  const pos = nearCity(i);
  return {
    lat: pos.lat, lng: pos.lng,
    altitude: rand(10, 1800),
    accuracy: rand(3, 15),
    heading: rand(0, 359),
    timestamp: ago(rand(0, 5)),
  };
}

function makeSensors(status: Vehicle["status"]): SensorData {
  const moving = status === "moving";
  return {
    speedKph:    moving ? rand(20, 120) : 0,
    fuelPercent: rand(8, 98),
    engineTempC: moving ? rand(82, 105) : rand(20, 50),
    engineOn:    status !== "offline" && status !== "stopped",
    odometreKm:  rand(5000, 250000),
    rpm:         moving ? rand(1200, 3800) : rand(0, 900),
    batteryV:    randF(11.8, 14.4),
    doorOpen:    Math.random() < 0.05,
    ignitionOn:  status === "moving" || status === "idle",
  };
}

const MAKES = [
  { make: "Toyota",  models: ["Hilux", "Land Cruiser", "Fortuner", "Quantum"] },
  { make: "Ford",    models: ["Ranger", "Transit", "Everest"] },
  { make: "Isuzu",   models: ["D-Max", "NLR 150", "NPR"] },
  { make: "VW",      models: ["Transporter", "Amarok", "Crafter"] },
  { make: "Mercedes",models: ["Sprinter", "Vito", "Actros"] },
  { make: "Hino",    models: ["300", "500", "700"] },
];

const FLEETS = ["Delivery", "Executive", "Field Ops", "Security", "Maintenance"];
const COLOURS = ["White", "Silver", "Black", "Blue", "Red", "Orange"];

const STATUSES: Vehicle["status"][] = ["moving", "moving", "moving", "idle", "idle", "stopped", "offline", "maintenance"];

// ─── Seed Data ────────────────────────────────────────────────────────────────

export const vehicleDb = {
  vehicles: Array.from({ length: 50 }, (_, i) => {
    const mk      = MAKES[i % MAKES.length];
    const model   = mk.models[i % mk.models.length];
    const status  = STATUSES[i % STATUSES.length];
    const pos     = makePosition(i);
    return {
      id:              uuid(),
      plateNumber:     `${["GP","WC","KZN","EC","FS","LP"][i % 6]} ${String(rand(100, 999))} ${String.fromCharCode(65 + (i % 26))}${String.fromCharCode(65 + ((i + 3) % 26))}`,
      make:            mk.make,
      model,
      year:            rand(2016, 2024),
      colour:          COLOURS[i % COLOURS.length],
      vin:             `VIN${String(i).padStart(14, "0")}`,
      fleetGroup:      FLEETS[i % FLEETS.length],
      driverId:        status !== "offline" ? `DRV-${String(i).padStart(3, "0")}` : null,
      status,
      commChannel:     (["cellular_4g","cellular_4g","cellular_4g","cellular_3g","satellite"][i % 5]) as Vehicle["commChannel"],
      position:        pos,
      sensors:         makeSensors(status),
      lastSeen:        ago(status === "offline" ? rand(30, 480) : rand(0, 3)),
      registeredAt:    ago(rand(1440, 525600)),
      simIccid:        `8927${String(i).padStart(15, "0")}`,
      deviceId:        `GT06N-${String(i).padStart(6, "0")}`,
      insuranceExpiry: new Date(Date.now() + rand(30, 365) * 86400000).toISOString(),
      licenceExpiry:   new Date(Date.now() + rand(60, 730) * 86400000).toISOString(),
      nextServiceKm:   rand(1000, 10000),
    } as Vehicle;
  }),

  drivers: Array.from({ length: 40 }, (_, i) => ({
    id:              `DRV-${String(i).padStart(3, "0")}`,
    name:            [
      "Sipho Nkosi","Thabo Dlamini","Lerato Mokoena","Nomsa Khumalo","David van der Berg",
      "Fatima Essop","Johan Botha","Priya Naidoo","Bongani Zulu","Elaine Mostert",
      "Khulekani Mthembu","Ayanda Cele","Pieter du Plessis","Zanele Mahlangu","Ravi Pillay",
      "Amina Mohamed","Chris Steyn","Nolwazi Dube","Stefan Kruger","Mmaphuti Mokgopa",
      "Lungelo Buthelezi","Sonia Ferreira","Thandeka Ntuli","Wayne Jacobs","Keabetswe Sithole",
      "Lebo Masondo","Mark Thompson","Zinhle Ndlovu","André Fourie","Busisiwe Mkhize",
      "Sanele Gumbi","Rebecca Jooste","Mthokozisi Ndaba","Charlene van Wyk","Nkosinathi Mhlanga",
      "Desiré Bosman","Vusi Shabalala","Lindi Pietersen","Siyanda Majola","Henk Swanepoel",
    ][i],
    licenceNo:       `SA${String(rand(1000000, 9999999))}`,
    phone:           `+2760${String(rand(1000000, 9999999))}`,
    email:           `driver${i}@vink-fleet.co.za`,
    status:          (i % 8 === 0 ? "suspended" : "active") as Driver["status"],
    assignedVehicleId: i < 35 ? null : null,
    totalTrips:      rand(50, 1200),
    totalKm:         rand(5000, 180000),
    safetyScore:     rand(55, 100),
    joinedAt:        ago(rand(1440, 525600)),
    avatarInitials:  "",
  } as Driver)).map(d => ({ ...d, avatarInitials: d.name.split(" ").map(p => p[0]).join("").slice(0, 2).toUpperCase() })),

  trips: Array.from({ length: 80 }, (_, i) => {
    const vehicleIdx = i % 50;
    const cityA = CITIES[i % CITIES.length];
    const cityB = CITIES[(i + 1) % CITIES.length];
    const start = ago(rand(30, 1440));
    const ended = i % 6 !== 0;
    const dist  = randF(5, 420);
    const avgSpd = rand(35, 95);
    const pts: import("../types/vehicles.js").TripPoint[] = Array.from({ length: 8 }, (_, p) => ({
      lat: +(cityA.lat + (cityB.lat - cityA.lat) * (p / 7) + randF(-0.02, 0.02)).toFixed(6),
      lng: +(cityA.lng + (cityB.lng - cityA.lng) * (p / 7) + randF(-0.02, 0.02)).toFixed(6),
      speedKph: rand(0, 130),
      timestamp: new Date(new Date(start).getTime() + p * rand(180000, 600000)).toISOString(),
    }));
    return {
      id:          uuid(),
      vehicleId:   `VEH-PLACEHOLDER-${vehicleIdx}`,
      driverId:    `DRV-${String(vehicleIdx % 40).padStart(3, "0")}`,
      startPos:    { lat: cityA.lat, lng: cityA.lng, altitude: 1200, accuracy: 5, heading: rand(0, 359), timestamp: start },
      endPos:      ended ? { lat: cityB.lat, lng: cityB.lng, altitude: 1100, accuracy: 5, heading: rand(0, 359), timestamp: ago(rand(0, 120)) } : null,
      startTime:   start,
      endTime:     ended ? ago(rand(0, 120)) : null,
      distanceKm:  dist,
      avgSpeedKph: avgSpd,
      maxSpeedKph: Math.min(rand(avgSpd + 10, 160), 160),
      fuelUsedL:   +(dist * randF(0.06, 0.12)).toFixed(2),
      points:      pts,
      ongoing:     !ended,
    } as Trip;
  }),

  alerts: Array.from({ length: 30 }, (_, i) => {
    const types: VehicleAlert["type"][] = ["speeding","geofence_breach","low_fuel","harsh_braking","engine_off","sos","tampering","accident"];
    const type = types[i % types.length];
    const sev: VehicleAlert["severity"] = type === "sos" || type === "accident" ? "critical" : type === "speeding" || type === "tampering" ? "warning" : "info";
    const pos = makePosition(i);
    return {
      id:           uuid(),
      vehicleId:    `VEH-PLACEHOLDER-${i % 50}`,
      plateNumber:  `GP ${rand(100,999)} AB`,
      driverId:     `DRV-${String(i % 40).padStart(3, "0")}`,
      type,
      severity:     sev,
      message: {
        speeding:        `Vehicle exceeded speed limit — ${rand(80,160)} km/h in 60 zone`,
        geofence_breach: `Vehicle exited authorised geofence boundary`,
        low_fuel:        `Fuel level critically low — ${rand(5,15)}% remaining`,
        harsh_braking:   `Harsh braking event detected at ${rand(60,110)} km/h`,
        engine_off:      `Engine switched off outside designated area`,
        sos:             `SOS panic button activated by driver`,
        tampering:       `GPS device tampering detected`,
        accident:        `Possible accident — sudden deceleration to 0 km/h`,
      }[type],
      position:     pos,
      detectedAt:   ago(rand(0, 480)),
      resolvedAt:   i % 4 === 0 ? ago(rand(0, 60)) : null,
      acknowledged: i % 3 === 0,
      value:        type === "speeding" ? rand(80, 160) : type === "low_fuel" ? rand(5, 15) : null,
    } as VehicleAlert;
  }),

  geofences: [
    { id: uuid(), name: "Johannesburg HQ",     description: "Head office compound",  shape: "circle", centerLat: -26.2041, centerLng: 28.0473, radiusM: 500,  active: true,  triggerOn: "exit",  alertVehicleGroups: ["Delivery","Executive"], createdAt: ago(43800), breachCount: 12 },
    { id: uuid(), name: "Cape Town Depot",      description: "Western Cape warehouse", shape: "circle", centerLat: -33.9249, centerLng: 18.4241, radiusM: 800,  active: true,  triggerOn: "both",  alertVehicleGroups: ["Delivery"],             createdAt: ago(20160), breachCount: 7  },
    { id: uuid(), name: "Durban Port Zone",     description: "Port operations area",  shape: "circle", centerLat: -29.8587, centerLng: 31.0218, radiusM: 1200, active: true,  triggerOn: "enter", alertVehicleGroups: ["Field Ops","Delivery"],  createdAt: ago(10080), breachCount: 3  },
    { id: uuid(), name: "Pretoria Admin Zone",  description: "Government precinct",   shape: "circle", centerLat: -25.7479, centerLng: 28.2293, radiusM: 600,  active: true,  triggerOn: "exit",  alertVehicleGroups: ["Executive"],            createdAt: ago(5040),  breachCount: 2  },
    { id: uuid(), name: "Night Curfew Area",    description: "After-hours restricted",shape: "circle", centerLat: -26.1052, centerLng: 28.0560, radiusM: 3000, active: false, triggerOn: "enter", alertVehicleGroups: ["Delivery","Field Ops"],  createdAt: ago(2880),  breachCount: 0  },
  ] as Geofence[],

  fleetHistory: Array.from({ length: 24 }, (_, i) => ({
    timestamp:       ago((23 - i) * 60),
    totalVehicles:   50,
    movingNow:       rand(18, 32),
    idleNow:         rand(5, 12),
    offlineNow:      rand(2, 6),
    maintenanceNow:  rand(1, 4),
    activeAlerts:    rand(3, 18),
    avgSpeedKph:     rand(45, 78),
    avgFuelPercent:  rand(52, 72),
    totalKmToday:    rand(800, 4200),
    activeDrivers:   rand(20, 35),
    geofenceBreaches: rand(0, 5),
  })) as FleetSnapshot[],
};

// Assign vehicle ids to trips
vehicleDb.trips.forEach((trip, i) => {
  trip.vehicleId = vehicleDb.vehicles[i % vehicleDb.vehicles.length].id;
});

// Assign alerts to real vehicle ids
vehicleDb.alerts.forEach((alert, i) => {
  const v = vehicleDb.vehicles[i % vehicleDb.vehicles.length];
  alert.vehicleId    = v.id;
  alert.plateNumber  = v.plateNumber;
  alert.driverId     = v.driverId;
});

export function getFleetSnapshot(): FleetSnapshot {
  const vv = vehicleDb.vehicles;
  return {
    timestamp:        new Date().toISOString(),
    totalVehicles:    vv.length,
    movingNow:        vv.filter(v => v.status === "moving").length,
    idleNow:          vv.filter(v => v.status === "idle").length,
    offlineNow:       vv.filter(v => v.status === "offline").length,
    maintenanceNow:   vv.filter(v => v.status === "maintenance").length,
    activeAlerts:     vehicleDb.alerts.filter(a => !a.resolvedAt).length,
    avgSpeedKph:      +( vv.filter(v => v.status === "moving").reduce((s, v) => s + v.sensors.speedKph, 0) / Math.max(1, vv.filter(v => v.status === "moving").length) ).toFixed(1),
    avgFuelPercent:   +( vv.reduce((s, v) => s + v.sensors.fuelPercent, 0) / vv.length ).toFixed(1),
    totalKmToday:     +( vehicleDb.trips.filter(t => t.startTime > ago(1440)).reduce((s, t) => s + t.distanceKm, 0) ).toFixed(1),
    activeDrivers:    vehicleDb.drivers.filter(d => d.status === "active" && vv.find(v => v.driverId === d.id && v.status === "moving")).length,
    geofenceBreaches: vehicleDb.alerts.filter(a => a.type === "geofence_breach" && !a.resolvedAt).length,
  };
}
